File: LeetCode/120_Triangle.py
# https://leetcode.com/problems/triangle/
# 2022年07月12日 14:42:45

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        # v1 DP
        # dp list recording the minimum of each path till last row
        dp=triangle[0].copy()
        for row in triangle[1:]: # do not include the first line cause it's the inital DP condition
            for idx in range(len(row)):
                if idx==0: # in the beginning of each row
                    row[idx] += dp[idx] # only upper right val can be selected
                elif idx==len(row)-1: # in the end of each row
                    row[idx] += dp[idx-1] # only upper left val can be selected
                else: # in the middle
                    row[idx] += min(dp[idx],dp[idx-1]) 
                    # choose the min between upper left and upper right val in dp list of last row
            dp=row.copy() # update dp
        return min(dp)

File: LeetCode/120_Triangle.java
// https://leetcode.com/problems/triangle/
// 2022年11月09日 17:23:09
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        for (int i = 1; i < triangle.size(); i++) {
            int rowSize = triangle.get(i).size();
            triangle.get(i).set(0, triangle.get(i).get(0) + triangle.get(i - 1).get(0));
            for (int j = 1; j < rowSize - 1; j++) {
                int minOfLast2 = Math.min(triangle.get(i - 1).get(j - 1), triangle.get(i - 1).get(j));
                triangle.get(i).set(j, triangle.get(i).get(j) + minOfLast2);
            }
            triangle.get(i).set(rowSize - 1, triangle.get(i).get(rowSize - 1) + triangle.get(i - 1).get(rowSize - 2));
        }
        int minOfLastRow = triangle.get(triangle.size() - 1).get(0);
        for (int val : triangle.get(triangle.size() - 1)) {
            minOfLastRow = Math.min(minOfLastRow, val);
        }
        System.out.println(triangle);
        return minOfLastRow;
    }
}